File: python/series de tiempo.py
# -*- coding: utf-8 -*-
"""
Created on Fri Aug 27 14:12:36 2021

@author: atita
"""

# Series de tiempo

# Cargar las siguientes librerias 

import pandas as pd # Es para leer los datos 
import numpy as np  # Es para arreglos de matrices y operaciones matematicas
import matplotlib.pyplot as plt # Es para la generacion graficos 

data= pd.read_excel("C:/Users/atita/Downloads/DATOS QUINUA.xlsx",header=1)

data.columns # Es para ver el nombre de las columnas 

data.head()

quinua= data["Producción"]

quinua.plot()

a= quinua.values

a=a.reshape(len(quinua),1)

data= pd.DataFrame(a)

data.columns= ["Produccion"]

# Los valores ausentes

data_new=data.replace(0,np.nan)

data_new.head()

n= len(data)
data_new.index= pd.date_range(start="2001-01-01",periods=n,freq="M")

data_new.plot()

# El metodo spline:

data_new= data_new.interpolate()

data_new.head()

# El metodo de holt-wintter

from statsmodels.tsa.seasonal import seasonal_decompose # trabajo con series de tiempo

dc =seasonal_decompose(data_new)

dc.plot()

from statsmodels.tsa.holtwinters import ExponentialSmoothing

data_new['TESadd12'] = ExponentialSmoothing(data_new,trend='add',seasonal='add',seasonal_periods=12).fit().fittedvalues
data_new.plot(figsize=(10,8))

modelo = ExponentialSmoothing(data_new["Produccion"],trend='add',seasonal='add',seasonal_periods=12).fit()

prediction=modelo.forecast(36)

data_new['Produccion'].plot(figsize=(12,8))
prediction.plot(label='Forecast additive model');
plt.legend(loc='upper left')

File: R/Series de tiempo.R
# Modelos de series de tiempo 


# Es ta funcion permite cargar y descargar los paquetes de forma automatica

ipak <- function(pkg){
  new.pkg <- pkg[!(pkg %in% installed.packages()[, "Package"])]
  if (length(new.pkg)) 
    install.packages(new.pkg, dependencies = TRUE)
  sapply(pkg, require, character.only = TRUE)
}

packages <- c("forecast","TSstudio","tseries","highcharter","TTR","readxl","imputeTS")
ipak(packages)

DATOS_QUINUA <- read_excel("C:/Users/atita/Downloads/DATOS QUINUA.xlsx")
View(DATOS_QUINUA)

# Ejemplo de valores Vacios o NA

a<- c(0,4,3) # Guardar en una variable 

mean(a)

a<- c(NA,4,3) # Guardar en una variable

mean(a,na.rm = TRUE)

#### 2 paso ####

data<- DATOS_QUINUA

print(data)

# Como veo si mi data tiene valores  vacios:

is.na(data)

sum(is.na(data))

# Valores ausentes por columnas 

colSums(is.na(data))

# Cambiar el valor por un valor ausente (NA)

data$Producción==0 # que valores son iguales a 0 

## Recordar: que cuando quiero comparar dos valores o objetos uso esto "==="

sum(data$Producción==0)

# Reamplazo: 

data$Producción[data$Producción==0]<- NA

colSums(is.na(data))

help("imputeTS")

#### 3 paso ####

# Vamos a darle el formato de serie de tiempo 

quinua_ts<- ts(data$Producción,start = c(2000,1),frequency = 12) # Transformar la data en
                                                                # Formato de serie de tiempo


plot.ts(quinua_ts) # Graficamos la serie

# Interpolar valores

new_data<- na.interpolation(quinua_ts,option = "linear") # Aplicamos el metodo de interpolacion

# Ojo: El metodo de interpolacion cubica o spline arroja valores negativos, por este motivos usamos el metodo
# lineal u linear 

class(new_data)

# hacemos la comparacion de las dos bases de datos 

plot(new_data) # con interpolacion 
lines(quinua_ts,col="red") # sin ella

plot(decompose(new_data))

ts_decompose(new_data)

ts_seasonal(new_data, type = "all")

#### 4 Aplicar el metodod de holt-winter ####

etsmodel = ets(new_data)

plot(new_data,col="blue")
lines(etsmodel$fitted,col="red")

# Multiplicativa

etsmodmult = ets(new_data, model ="MMM")

plot(new_data,col="blue")
lines(etsmodmult$fitted,col="red")

autoplot(etsmodmult)

# Realizar una prediccion

prediction <- forecast(etsmodmult,h=48)

autoplot(prediction)


